### ISSUE TITLE
Using example code results in mypy complaining about `Type[SearchRequest]` has no attribute `loads` nor `dumps`

### ISSUE THREAD

**Issue by giuliano-oliveira** [eigenein/protobuf#102]
Firstly, great job with this package I really disliked using vanilla protobuf poor type hinted code, using dataclasses makes my life wayy easier.

## Issue
Using the example code in the readme makes mypy complain about missing `loads` and `dumps` method in `Type[SearchRequest]`, even though the code is functional and both methods exists in runtime, example code:
```python
from dataclasses import dataclass

from pure_protobuf.dataclasses_ import field, message
from pure_protobuf.types import int32


@message
@dataclass
class SearchRequest:
    query: str = field(1, default="")
    page_number: int32 = field(2, default=int32(0))
    result_per_page: int32 = field(3, default=int32(0))


obj = SearchRequest(
    query="hello",
    page_number=int32(1),
    result_per_page=int32(10),
)
serialized = obj.dumps()
deserialized = SearchRequest.loads(serialized)
print(f"{serialized=}")
print(f"{deserialized=}")
```

output:
```
serialized=b'\n\x05hello\x10\x01\x18\n'
deserialized=SearchRequest(query='hello', page_number=1, result_per_page=10)
```

mypy output:
```
example.py
example.py:3: error: Skipping analyzing "pure_protobuf.dataclasses_": module is installed, but missing library stubs or py.typed marker  [import]
example.py:3: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
example.py:4: error: Skipping analyzing "pure_protobuf.types": module is installed, but missing library stubs or py.typed marker  [import]
example.py:20: error: "SearchRequest" has no attribute "dumps"  [attr-defined]
example.py:21: error: "Type[SearchRequest]" has no attribute "loads"  [attr-defined]
Found 4 errors in 1 file (checked 1 source file)
```

## workaround
Though a little bit repetitive, it is possible to use the base class super method + `getattr` to explicit define a dumps and loads method:
```python
@message
@dataclass
class SearchRequest:
    query: str = field(1, default="")
    page_number: int32 = field(2, default=int32(0))
    result_per_page: int32 = field(3, default=int32(0))

    def dumps(self) -> bytes:
        return getattr(super(), "dumps")()

    @classmethod
    def loads(cls, data: bytes) -> "SearchRequest":
        return getattr(super(), "loads")(data)
```

## Versions info
* python: 3.10.8
* mypy: 0.991 (compiled: yes)
* pure-protobuf: 2.2.0
* os: Linux 5.15.85-1-MANJARO

**Comment by eigenein** [comment]
Hi Giuliano,

Thanks for the feedback and sponsorship!

There's a bad news and good news to that:

### The bad news

The current code contains a lot of legacies from the epoch of Python 2.x. Even the existing type hints is a relatively recent addition.

Technically, `@message` [registers](https://docs.python.org/3/library/abc.html#abc.ABCMeta.register) this child class as a «virtual subclass» of `Message` and patches it, but surely that's not enough for `mypy` and/or IDEs.

I'd go for `# type: ignore` for the `loads` and `cast(Message, obj).dumps()` to work around that.

I'm not quite sure, if it's worth putting effort in a permanent solution for that because there's…

### The good news

I've been working on the major 3.0 revamp to re-write the legacies: [`milestone/v3`](https://github.com/eigenein/protobuf/tree/milestone/v3). It's not merged yet, but it already passes the tests and type linters (just there's still a couple of improvements that I'd like to address).

In particular, this issue is solved there by simply inheriting from `BaseMessage`, which naturally provides the relevant methods. No patching any more, and the `= field(…)` assignment is replaced with `Annotated`. Way more typing-friendly.

Feel free to take a look at the [`README`](https://github.com/eigenein/protobuf/blob/milestone/v3/README.md) and open a discussion.

**Comment by giuliano-oliveira** [comment]
Thanks for the detailed answer! It is very clear to me what is the issue!

It is OK if I leave this issue open while V3 isn't merged, or should I close it?

**Comment by eigenein** [comment]
It's alright to keep it, I've just tagged it with the milestone

**Comment by Mokto** [comment]
In the meantime, I've created a subclass like this:

```
from dataclasses import dataclass

from pure_protobuf.dataclasses_ import message
from typing_extensions import Self


@message
@dataclass
class ProtoBase:
    def dumps(self) -> bytes:
        return getattr(super(), "dumps")()

    @classmethod
    def loads(cls, data: bytes) -> Self:
        return getattr(super(), "loads")(data)
```

You can then use it like this:

```
@message
@dataclass
class MyClass(ProtoBase):
    website: str = field(1, default="")
```

Hope this helps!

**Comment by eigenein** [comment]
Should be fixed in `3.0.0`.
